dbms/src/Functions/FunctionsString.cpp

```diff
@@ -4392,6 +4392,109 @@ class FunctionRepeat : public IFunction
+
+    FunctionSpace() = default;
+
+    static FunctionPtr create(const Context & /*context*/)
+    {
+        return std::make_shared<FunctionSpace>();
+    }
+
+    std::string getName() const override { return name; }
+    size_t getNumberOfArguments() const override { return 1; }
+
+    bool useDefaultImplementationForConstants() const override { return true; }
+
+    DataTypePtr getReturnTypeImpl(const DataTypes & arguments) const override
+    {
+        return arguments[0]->onlyNull()
+            ? makeNullable(std::make_shared<DataTypeNothing>())
+            : makeNullable(std::make_shared<DataTypeString>());
+    }
+
+    void executeImpl(Block & block, const ColumnNumbers & arguments, size_t result) const override
+    {
+        if (executeSpace<UInt8>(block, arguments, result)
+            || executeSpace<UInt16>(block, arguments, result)
+            || executeSpace<UInt32>(block, arguments, result)
+            || executeSpace<UInt64>(block, arguments, result)
+            || executeSpace<Int8>(block, arguments, result)
+            || executeSpace<Int16>(block, arguments, result)
+            || executeSpace<Int32>(block, arguments, result)
+            || executeSpace<Int64>(block, arguments, result))
+        {
+            return;
+        }
+        else
+        {
+            throw Exception(fmt::format("Illegal argument of function {}", getName()), ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT);
+        }
+    }
+
+private:
+    template <typename IntType>
+    bool executeSpace(
+        Block & block,
+        const ColumnNumbers & arguments,
+        const size_t result) const
+    {
+        auto & c0_col = block.getByPosition(arguments[0]);
+
+        Field res_field;
+        auto c0_col_column = c0_col.column;
+        const auto * col_vector_space_num = checkAndGetColumn<ColumnVector<IntType>>(c0_col_column.get());
+        if (col_vector_space_num == nullptr)
+        {
+            return false;
+        }
+        const auto & col_vector_space_num_value = col_vector_space_num->getData();
+        size_t val_num = block.rows();
+        auto result_null_map = ColumnUInt8::create(val_num);
+        auto col_res = ColumnString::create();
+
+        auto & col_res_data = col_res->getChars();
+        auto & col_res_offsets = col_res->getOffsets();
+
+        col_res_offsets.resize(c0_col_column->size());
+        ColumnString::Offset res_offset = 0;
+
+
+        for (size_t row = 0; row < val_num; ++row)
+        {
+            result_null_map->getData()[row] = false;
+
+            Int64 space_num = accurate::lessOp(INT64_MAX, col_vector_space_num_value[row]) ? INT32_MAX : col_vector_space_num_value[row];
+            if (space_num < 0)
+            {
+                space_num = 0;
+            }
+            if (space_num > MAX)
+            {
+                result_null_map->getData()[row] = true;
+                space_num = 0;
+            }
+            col_res_data.resize(col_res_data.size() + space_num + 1);
```

Comment:
better to reserve some space before the loop.

Reply:
And also since in most of the user case, the argument of `space` should be a constant, so I think we should not set  `useDefaultImplementationForConstants` to true, and handle constant column explicitly.